tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php

```diff
@@ -281,7 +286,11 @@ public function testTokenRendering(): void {
       non_deductible_amount = {contribution.non_deductible_amount}
       total_amount = {contribution.total_amount}
       net_amount = {contribution.net_amount}
-      fee_amount = {contribution.fee_amount}';
+      fee_amount = {contribution.fee_amount}
+      legacy campaign = {contribution.campaign}
+      campaign_id = {contribution.campaign_id}
+      campaign name = {contribution.campaign_id:name}
```

Comment:
OK - so it looks like you added the pseudoconstant form - I think we should either make this work in apiv4 or use the dot notation per https://github.com/civicrm/civicrm-core/pull/21076/files#diff-196c289971f24f7589d89f613b6199310f614fbe5e2aac9368b1563bf64d2f96R291